File: algorithm/golang/weekcontest/wc363/countWays.go
package wc363

// 超时
func countWays(nums []int) int {
	n := len(nums)
	check := func(k int) bool {
		cnt := 0
		for _, num := range nums {
			if num == k {
				return false
			}
			if cnt > k {
				return false
			}
			if k > num {
				cnt++
			}
		}
		return cnt == k
	}
	res := 0
	for k := 0; k <= n; k++ {
		if check(k) {
			res++
		}
	}
	return res
}

File: algorithm/clang/src/wc363/countWays.c
//
// Created by leiwingqueen on 2023/9/17.
//
#include "stdio.h"
#include "stdlib.h"

char check(int *nums, int n, int k) {
    int l = 0, r = n - 1;
    while (l < r) {
        int mid = l + (r - l) / 2;
        if (nums[mid] > k) {
            r = mid;
        } else {
            l = mid + 1;
        }
    }
    if (nums[l] <= k) {
        l = n;
    }
    char res = l == k && (l == 0 || nums[l - 1] != k);
    // printf("k:%d,idx:%d,res:%d\n", k, l, res);
    return res;
}

int cmp(const void *a, const void *b) {
    return *(int *) a - *(int *) b;//升序
}

int countWays(int *nums, int numsSize) {
    int res = 0;
    qsort(nums, numsSize, sizeof(int), cmp);
    for (int i = 0; i <= numsSize; ++i) {
        if (check(nums, numsSize, i)) {
            res++;
        }
    }
    return res;
}

